File: 111. Minimum Depth of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <climits>
class Solution {
private:
    int minheight = INT_MAX;
public:
    int minDepth(TreeNode* root) {
        if (root == NULL) return 0;
        DFS(root, 0);
        return minheight;
    }
    
    void DFS(TreeNode* root, int height)
    {
        if (root == NULL) return;
        else if (root->left == NULL && root->right == NULL)
        {
            if (height + 1 < minheight) minheight = height + 1;
            return;
        }
        DFS(root->left, height + 1);
        DFS(root->right, height + 1);
    }
};

File: 111. Minimum Depth of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import sys
class Solution(object):
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        self.minheight = sys.maxint
        self.DFS(root, 0)
        return self.minheight
        
    def DFS(self, root, height):
        if not root:
            return
        if not root.left and not root.right:
            if height + 1 < self.minheight:
                self.minheight = height + 1
            return
        height += 1
        self.DFS(root.left, height)
        self.DFS(root.right, height)
